src/backports/three_way_comparison.h

```diff
@@ -0,0 +1,384 @@
+
+template <typename LT, typename RT>
+constexpr bool hasExternalCompareThreeWay_v =
+    HasExternalCompareThreeWay<LT, RT>::value;
+template <typename LT, typename RT>
+constexpr bool hasMemberCompareThreeWay_v =
+    HasMemberCompareThreeWay<LT, RT>::value;
+
+template <typename LT, typename RT>
+struct HasAnyCompareThreeWay
+    : std::disjunction<HasExternalCompareThreeWay<LT, RT>,
+                       HasMemberCompareThreeWay<LT, RT>> {};
+
+template <typename LT, typename RT>
+constexpr bool hasAnyCompareThreeWay_v = HasAnyCompareThreeWay<LT, RT>::value;
+
+template <typename LT, typename RT>
+constexpr std::enable_if_t<
+    std::conjunction_v<std::is_floating_point<LT>, std::is_floating_point<RT>>,
+    ql::partial_ordering>
+compareThreeWay(const LT& lhs, const RT& rhs) {
+  if (std::isnan(lhs) || std::isnan(rhs)) {
+    return ql::partial_ordering::unordered;
+  }
+  if (lhs < rhs) {
+    return ql::partial_ordering::less;
+  }
+  if (lhs > rhs) {
+    return ql::partial_ordering::greater;
+  }
+  return ql::partial_ordering::equivalent;
+}
+
+template <typename LT, typename RT,
+          std::enable_if_t<std::conjunction_v<HasAllComparisonOperators<LT, RT>,
+                                              std::negation<std::disjunction<
+                                                  HasAnyCompareThreeWay<LT, RT>,
+                                                  std::is_floating_point<LT>,
+                                                  std::is_floating_point<RT>>>>,
+                           int> = 0>
+constexpr auto compareThreeWay(const LT& lhs, const RT& rhs) {
+  if (lhs < rhs) {
+    return ql::strong_ordering::less;
+  } else if (lhs > rhs) {
+    return ql::strong_ordering::greater;
+  }
+  return ql::strong_ordering::equal;
+}
+
+template <typename LT, typename RT,
+          std::enable_if_t<ql::HasAnyCompareThreeWay<LT, RT>::value, int> = 0>
+constexpr auto compareThreeWay(const LT& lhs, const RT& rhs) {
+  if constexpr (hasMemberCompareThreeWay_v<LT, RT>) {
+    return lhs.compareThreeWay(rhs);
+  } else {
+    return compareThreeWay(lhs, rhs);
+  }
+}
+
+template <typename T, std::enable_if_t<std::is_integral_v<T>, int> = 0>
+constexpr auto compareThreeWay(const ql::strong_ordering& lhs,
+                               const T& /*rhs*/) {
+  // Return unchanged
+  return lhs;
+}
+
+template <typename T, std::enable_if_t<std::is_integral_v<T>, int> = 0>
+constexpr auto compareThreeWay(const T& /*lhs*/,
+                               const ql::strong_ordering& rhs) {
+  // Invert ordering
+  if (rhs < 0) {
+    return ql::strong_ordering::greater;
+  } else if (rhs > 0) {
+    return ql::strong_ordering::less;
+  }
+  return std::strong_ordering::equal;
+}
+
+}  // namespace ql
+
+#define QL_CONSTEXPR_IF(C) constexpr C
+
+#define QL_DEFINE_CLASS_MEMBERS_AS_TIE_(CONSTEXPR, ...) \
+  CONSTEXPR auto membersAsTie() const { return std::make_tuple(__VA_ARGS__); }
```

Comment:
This should definitely use `std::tie`, because we want references and not deep copies here.